### PR TITLE
feat: expand source jsonschema

### PR DESCRIPTION
**Pull request by howbazaar**
Allow the source definition to be an object, with the name being the type, and other attributes up to the source itself.

This will allow attributes of the source to be kept with the source rather than the top level `query` policy attribute.

### CODE REVIEW

File: c7n/query.py [hunk 1]
```diff
@@ -475,11 +475,22 @@ def __init__(self, ctx, data):
...
+
+    @property
+    def source_config(self):
+        value = self.data.get('source', 'describe')
+        return value if isinstance(value, dict) else {}
 
     def get_source(self, source_type):
         if source_type in self.source_mapping:
-            return self.source_mapping.get(source_type)(self)
+            return self.source_mapping[source_type](self)
```
[the remark this answers is not in the document]
Comment (ajkerrigan) [reply c7n/query.py:493]: Yes, `sources` is the set of sources registered in the general `QueryResourceManager`, while `source_mapping` might be overridden at the resource level. For one concrete example, peeking at what `get_sources()` sees when running an EC2 policy...

```python
sources['describe']
# <class 'c7n.query.DescribeSource'>

self.source_mapping['describe']
# <class 'c7n.resources.ec2.DescribeEC2'>
```

File: c7n/query.py [hunk 2]
```diff
@@ -475,11 +475,22 @@ def __init__(self, ctx, data):
...
+        value = self.data.get('source', 'describe')
+        if isinstance(value, str):
+            return value
+        # Here we assume that the data has been validated, so the other
+        # alternative is that the source is a dict with the name attribute
+        # being a valid source type.
+        return value["name"]
+
+    @property
+    def source_config(self):
```
Comment (ajkerrigan) [inline c7n/query.py:487]: It does seem useful to have a separate property for the configuration, though I see some potential confusion between `source_config` and the config source. Definitely a nit, and possibly unavoidable.

Still it's possible that something like `source_settings` or `source_attributes` could also work? :shrug:

**Review by ajkerrigan — COMMENTED** [review]
I dig this generally. It's been an occasional point of confusion that a top-level `query` block can mean both "server-side filters" (for the `describe` source) or "advanced SQL queries" (for the `config` source).

Might be worth deprecating/advertising this change, and trying to help folks migrate from top-level `query` blocks to source-specific config over time. Thanks for this.

**Comment by ajkerrigan** [comment]
FWIW, failing checks are related to some API changes related to the awscc provider. A change like [this](https://github.com/cloud-custodian/cloud-custodian/pull/9161#issuecomment-1819892998) seems like one way to handle it, but it may not be the best.

**Comment by ajkerrigan** [comment]
> FWIW, failing checks stem from some API changes related to the awscc provider. A change like [this](https://github.com/cloud-custodian/cloud-custodian/pull/9161#issuecomment-1819892998) seems like one way to handle it, but it may not be the best.

#9165 should take care of that failing check.

**Comment by howbazaar** [comment]
Closed in favour of #9167

**Pull request closed without merging**